search/search_repository.go

```diff
@@ -382,11 +379,70 @@ func (q Query) determineLiteralType(key string, val string) criteria.Expression
 func (q Query) generateExpression() (criteria.Expression, error) {
 	var myexpr []criteria.Expression
 	currentOperator := q.Name
-	if !isOperator(currentOperator) {
-		var key string
-		if val, ok := searchKeyMap[q.Name]; ok {
-			key = val
+
+	// Here we handle the "hierarchy" query parameter which we translate from a
+	// simple
+	//
+	// "hierarchy = y"
+	//
+	// expression into an
+	//
+	// "Type in (y1, y2, y3, ... ,yn)"
+	//
+	// expression where yi represents the i-th work item type associated with
+	// the hierarchy y.
+	handleHierarchy := func(q Query) error {
+		if q.Value == nil {
+			return errors.NewBadParameterError(hierarchy, q.Value).Expected("not nil")
+		}
+		typeGroupMap := map[string]*typegroup.WorkItemTypeGroup{
+			typegroup.Execution0.BuildName():    &typegroup.Execution0,
+			typegroup.Portfolio0.BuildName():    &typegroup.Portfolio0,
+			typegroup.Portfolio1.BuildName():    &typegroup.Portfolio1,
+			typegroup.Requirements0.BuildName(): &typegroup.Requirements0,
+		}
+		typeGroup, ok := typeGroupMap[*q.Value]
```

Comment:
should API use `BuildNames` in query payload? like `Portfolio-0.1` `Portfolio-1.0` etc?
How about using different `search_key` for every type group?

Reply:
@pranavgore09 sure, that would work but for now I didn't want to change the underlying data-model. Changing to a different field is a piece of cake.